zerver/worker/thumbnail.py

```diff
@@ -120,9 +127,46 @@ def ensure_thumbnails(image_attachment: ImageAttachment) -> int:
             # We have never thumbnailed this -- it most likely had
             # bad data.  Remove the ImageAttachment row, since it is
             # not valid for thumbnailing.
+            update_message_rendered_content(
+                image_attachment.realm_id, image_attachment.path_id, None
+            )
             image_attachment.delete()
+            # TODO: Clean up any dangling thumbnails we may have
+            # produced?  Seems unlikely that we'd fail on one size,
+            # but not another, but anything's possible.
             return 0
 
     image_attachment.save(update_fields=["thumbnail_metadata"])
-
+    update_message_rendered_content(
+        image_attachment.realm_id,
+        image_attachment.path_id,
+        get_default_thumbnail_url(image_attachment),
+    )
     return written_images
+
+
+def update_message_rendered_content(
+    realm_id: int, path_id: str, image_data: tuple[str, bool] | None
+) -> None:
+    for message_class in [Message, ArchivedMessage]:
+        messages_with_image = (
+            message_class.objects.filter(  # type: ignore[attr-defined]  # TODO: ?
+                realm_id=realm_id, attachment__path_id=path_id
+            )
+            .select_for_update()
+            .order_by("id")
+        )
+        for message in messages_with_image:
+            rendered_content = rewrite_thumbnailed_images(
+                message.rendered_content,
+                {} if image_data is None else {path_id: image_data},
+                {path_id} if image_data is None else set(),
+            )
+            if rendered_content is None:
+                continue
```

Comment:
What do we expect this circumstance to reflect?

Reply:
We didn't update the content.  This happens if we re-thumbnail the same image for some reason, and the spinner isn't in there anymore to be replaced.

Reply:
Yeah, that's clear now having read the rest of the PR, but probably worth adding a comment explaining the meaning for general readability, since we're using `None` as a special return value.

Reply:
Added a comment